src/main/scala/com/microsoft/hyperspace/util/ResolverUtils.scala

```diff
@@ -16,19 +16,73 @@
 
 package com.microsoft.hyperspace.util
 
-import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.{Column, SparkSession}
 import org.apache.spark.sql.catalyst.analysis.Resolver
 import org.apache.spark.sql.catalyst.expressions.{Alias, Attribute, Expression, ExtractValue, GetArrayStructFields, GetMapValue, GetStructField}
 import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.functions.col
 import org.apache.spark.sql.types.{ArrayType, MapType, StructField, StructType}
 
 import com.microsoft.hyperspace.HyperspaceException
+import com.microsoft.hyperspace.util.ResolverUtils.ResolvedColumn.NESTED_FIELD_PREFIX
 
 /**
  * [[ResolverUtils]] provides utility functions to resolve strings based on spark's resolver.
  */
 object ResolverUtils {
 
+  /**
+   * [[ResolvedColumn]] stores information when a column name is resolved against the
+   * analyzed plan and its schema.
+   *
+   * Outside unit tests, this object should not be created directly, but via the `resolve` function,
+   * or `ResolvedColumn.apply` with a normalized name.
+   *
+   * @param name The column name resolved from an analyzed plan.
+   * @param isNested Flag to denote if this column is nested or not.
+   */
+  private[hyperspace] case class ResolvedColumn(name: String, isNested: Boolean) {
+    assert(!isNested || (name.contains(".") && !name.startsWith(NESTED_FIELD_PREFIX)))
+
+    // For nested fields, the column name is prefixed with `NESTED_FIELD_PREFIX`.
+    lazy val normalizedName = {
+      if (isNested) {
+        s"$NESTED_FIELD_PREFIX$name"
+      } else {
+        name
+      }
+    }
+
+    // Create a column using the resolved name. Top level column names are quoted, and
+    // nested column names are aliased with normalized names.
+    def toColumn: Column = {
+      if (isNested) {
+        // No need to quote the string for "as" even if it contains dots.
+        col(name).as(normalizedName)
+      } else {
+        col(quote(name))
+      }
+    }
+
+    // Create a column using the normalized name. Since the normalized name is already flattened
+    // with "dots", it is quoted.
+    def toNormalizedColumn: Column = col(quote(normalizedName))
+
+    private def quote(name: String) = s"`$name`"
+  }
+
+  private[hyperspace] object ResolvedColumn {
+    private val NESTED_FIELD_PREFIX = "__hs_nested."
+
+    def apply(normalizedColumnName: String): ResolvedColumn = {
```

Comment:
Can we add some UTs to verify the `ResolvedColumn.apply` functionality? There are some conditions in `apply` method that should be covered by tests.

Reply:
`assert(col2 === ResolvedColumn(col2.normalizedName))`?